soc: arm: atmel_sam: Add sams70 soc

mcscholtz opened the pull request:
This patch adds support for the sams70 family, these chips are similar to the same70 but have less features

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>


*Note: This is blocked by the following [PR](https://github.com/zephyrproject-rtos/hal_atmel/pull/18) in the atmel hal repo. It will need to be merged as this PR depends on it.

Commit:
soc: arm: atmel_sam: Add sams70 soc

This patch adds support for the sams70 family,
these chips are similar to the same70 but have
less features.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

Commit:
drivers: gpio: atmel_sam: Add support for SAMS70

This patch adds support to the sam gpio driver for
the SAMS70 soc.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

Commit:
drivers: counter: atmel_sam: Add support for SAMS70 soc

This patch adds support for the SAMS70 soc to the
sam counter driver.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

nandojve commented:
Hi @mcscholtz ,

As general police in Zephyr, recently new SoC code is only integrated when there is a board that can be compiled, just to ensure that code at least compile. In this case, this PR should have at least:

 * update hal_atmel (https://github.com/zephyrproject-rtos/hal_atmel/pull/18)
 * soc: arm: atmel_sam: Add sams70 soc
 * board: arm: some_sams70_board

mcscholtz commented:
@nandojve Okay, that might be a problem since I do not know of any publicly available boards with this soc. We are using it in a custom board.

Can I create a minimal generic dummy board for this purpose, even though there is no actual hardware for it?

stephanosio commented:
> Can I create a minimal generic dummy board for this purpose, even though there is no actual hardware for it?

It looks like SAM S70 is strictly a subset of the SAM E70, so you should be able to use SAM E70 Xplained board to run the code targeting SAM S70.

stephanosio commented on dts/arm/atmel/sams70.dtsi:
```diff
@@ -0,0 +1,422 @@
+
+#include "sams70-pinctrl.dtsi"
+
+/ {
```
Since SAM S70 is supposed to be a subset of the already supported SAM E70 series, the SAM S70 dtsi should be made a "base" dts, and the SAM E70 dtsi should be updated to include the SAM S70 dtsi for the sake of future maintainability.

We are doing the same thing for SAM V71, which is a superset of the SAM E70 series:
https://github.com/zephyrproject-rtos/zephyr/blob/284d2a11db131198a19f2b189e42fa0d5e925064/dts/arm/atmel/samv71.dtsi#L7

Note that the SAM V71 dtsi includes the SAM E70 dtsi instead of unnecessarily duplicating all the information.

mcscholtz replied:
Agreed, I will change this. There is a similar situation with the pincontrol files that I will fix also.

mcscholtz replied:
@stephanosio I have refactored the dts files to make it more maintainable. Please have a look when you have time. Once this is okay I will go ahead and create a board.

nandojve replied:
Hi @mcscholtz , by @stephanosio suggestion, you don't need create a new board. You just need add another board target inside sam_e70_xplained directory. In fact, this board already have two targets `sam_e70_xplained` and `sam_e70b_xplained`.

mcscholtz replied:
Thanks @nandojve and @stephanosio 

I have done as you suggested and added a separate target for the SAMS70 to the existing ```sam_e70_xplained``` board.

It behaves as expected and will compile ```zephyr/samples/basic/blinky``` but fails on the ```zephyr/samples/net/*``` since the S70 does not have ethernet.

Just a question about the dts files: 

I added all the peripherals supported by the S70 to the sams70.dtsi. Some peripherals are only supported depending on which package is used. 

For example:
spi1 is only available on Q packages but not on J or N packages. 

However spi1 is added to the sams70.dtsi file, so there should be a really be separate file for each package and the J one should be the base. 

I added comments to the pinctrl file to indicate which peripherals are only available to which packages. Is this something that can be changed in a different PR?

Also the build seems to be failing because the hal is missing

stephanosio replied:
> Also the build seems to be failing because the hal is missing

@mcscholtz Please update `west.yml` to point to the PR https://github.com/zephyrproject-rtos/hal_atmel/pull/18.

See https://docs.zephyrproject.org/latest/guides/modules.html#process-for-submitting-changes-to-existing-modules.

stephanosio reviewed (changes requested):
As @nandojve pointed out, we need at least one board (and tests/samples using it) before we can accept this.

My suggestion is to create a variant for the `sam_e70_xplained` board that "targets" SAM S70. It should be possible as long as the SAM S70 is strictly a subset of the SAM E70 series.

Commit:
dts: arm: atmel_sam: Refactor dts for SAM E70/V71/S70

This patch is to refactor the dts for the SAM E70/V71 to
include the SAM S70 in a maintainable way.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

Commit:
boards: arm: sam_e70_xplained: Added support for SAMS70

This patch adds a separate target for the sam_e70_xplained
board to target only the SAMS70 subset. The SAMS70 does not
support ethernet or can. The SAMS70 subset can be targeted
by compiling for the sam_s70_xplained or sam_s70b_xplained.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

mcscholtz commented:
Hi @stephanosio I have included the hal PR. It looks like the only failures are related to deprecated macro warnings in the hal see below:
```
/__w/zephyr/zephyr/drivers/serial/usart_sam.c: In function 'usart_sam_init':
/__w/zephyr/zephyr/drivers/serial/usart_sam.c:72:13: error: Using deprecated macro US_MR_NBSTOP_1_BIT [-Werror]
   72 |  usart->US_MR =   US_MR_NBSTOP_1_BIT
      |             ^~~~~~~~~~~~~~~~~~~~~~~~                   
/__w/zephyr/zephyr/drivers/serial/usart_sam.c:72:13: error: Using deprecated macro US_MR_NBSTOP_1_BIT_Val [-Werror]
   72 |  usart->US_MR =   US_MR_NBSTOP_1_BIT
      |             ^~~~~~~~~~~~~~~~~~~~~~~~                       
/__w/zephyr/zephyr/drivers/serial/usart_sam.c:72:13: error: Using deprecated macro US_MR_NBSTOP_Pos [-Werror]
   72 |  usart->US_MR =   US_MR_NBSTOP_1_BIT
      |             ^~~~~~~~~~~~~~~~~~~~~~~~                 
/__w/zephyr/zephyr/drivers/serial/usart_sam.c:73:13: error: Using deprecated macro US_MR_PAR_NO [-Werror]
   73 |          | US_MR_PAR_NO
      |             ^~~~~~~~~~~                          
```

Looks like there is a deprecated directory in the hal ```atmel/asf/sam/include/samx70b/component/deprecated``` but only for the b variants, including the existing sam e70b. There are deprecated entries for usart, usbhs and tc.

Seems to be a name change only, for example ```US_CR_DTRDIS``` becomes ```US_CR_USART_DTRDIS```

How do we want to fix this?

nandojve commented:
Hi @mcscholtz ,

In general we've been fixing SAM SoCs directly in the hal_atmel, case by case.

mcscholtz commented:
Thanks @nandojve,

Seems me trying to fix the deprecated warnings for the x7x have made the test fail for some of the other sam chips. Can I just change it back and we can ignore the deprecated warnings or disable the warnings?

Also the ethernet driver keeps trying to build for the S70 and I am not sure how to disable it. Any idea how to disable it for the S70?

nandojve commented:
> Thanks @nandojve,
> 
> Seems me trying to fix the deprecated warnings for the x7x have made the test fail for some of the other sam chips. Can I just change it back and we can ignore the deprecated warnings or disable the warnings?

Yes, change it back and update the patch at hal_atmel to use current U[S]ART symbols.

Commit:
soc: arm: atmel_sam: s70: Remove fixups in favour of hal fixes

The patch removes the soc fixups so that it can be done in the hal
repo.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

Commit:
boards: arm: sam_e70_xplained: Remove network options from s70

Remove the network options from the S70 board because it does not
have ethernet like the E70 and V71.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

Commit:
drivers: arm: flash: atmel_sam: Add S70 soc

Add the sam S70 soc in the KConfig to enable flash support
on the S70.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

nandojve commented:
Hi @mcscholtz ,

I'll look at this after #43731.
